### ISSUE TITLE
A smarter nrefine (/ nrapply / ...)

### ISSUE THREAD

**Issue by mikeshulman** [HoTT/HoTT#1241]
It would be pleasing to have a tactic that does `nrefine` and then applies typeclass search on a specified list of the generated subgoals.  So that something like

```coq
nrefine (some_lemma _ a _ b).
1-2,3,5:exact _.
```

could be done with one tactic.  Is there a way to do that?

**Comment by Alizter** [comment]
I don't think this is possible in ltac but perhaps in ltac2. @JasonGross what do you think?

**Comment by JasonGross** [comment]
What's your desired syntax? / where's an example where this would make life easier?

Maybe you could make a notation for `ltac:(typeclasses eauto)` and use that for some holes rather than `_`?

**Comment by mikeshulman** [comment]
Whoa, you can make a notation for `ltac:(...)` snippets?

I was thinking of something like `trefine (some_lemma _ a _ b) 1 2 3 5.`  Often the goals that typeclass resolution gets invoked on are not explicit `_` holes in the term.

**Comment by Alizter** [comment]
@JasonGross How can I make the following work then:

```coq
Require Import Basics.
Require Import Truncations.
Import TrM.

Notation "%" := ltac:(typeclasses eauto).

Section foo.
Context {A B C D} {f : A -> B} {g : B -> C} {h : C -> D}.
Context `{!IsConnMap 5 f, !IsConnMap 5 g, !IsConnMap 5 h}.

Goal IsConnMap 5 (h o g o f).

nrefine (@conn_map_compose _ _ _ _ _ _ % _).

```
?

**Comment by JasonGross** [comment]
> Whoa, you can make a notation for ltac:(...) snippets?

Yep!  But maybe they don't work in uconstr.

@Alizter Hmm, if that doesn't work, try marking the goal:
```coq
Inductive marked := mark.
Notation "%" := (let a := mark in _).
Tactic Notation "trefine" uconstr(lem)
  := nrefine lem;
     lazymatch goal with
     | [ H : marked |- _ ] => exact _
     | _ => idtac
     end.
```

@mikeshulman I don't think you can do that without writing an OCaml plugin.  I don't think tactic notations currently let you pass in goal selector syntax, and Ltac is quite bad at multigoal tactics.  You can probably do it with Ltac2, though.

**Comment by Alizter** [comment]
That seems to work, thanks @JasonGross.

@mikeshulman what do you think about:
```coq
Require Import Basics.
Require Import Truncations.
Import TrM.

Inductive marked := mark.
Notation "%" := (let a := mark in _).
Tactic Notation "trefine" uconstr(lem)
  := nrefine lem;
     lazymatch goal with
     | [ H : marked |- _ ] => exact _
     | _ => idtac
     end.

Section foo.
Context {A B C D} {f : A -> B} {g : B -> C} {h : C -> D}.
Context `{!IsConnMap 5 f, !IsConnMap 5 g, !IsConnMap 5 h}.

Goal IsConnMap 5 (h o g o f).

trefine (@conn_map_compose _ _ _ _ _ _ % _).
```

**Comment by mikeshulman** [comment]
That's not bad.  I'll have to experiment with it a bit to see how helpful it is in practice.

**Comment by jdchristensen** [comment]
Something short would be great, but I don't mind
```
nrefine (some_lemma _ a _ b); only 1-2,3,5: exact _.
```
or the form with a period that Mike posted at the top of this issue.  Each non-standard tactic will puzzle some newcomers, and if it's only needed here and there, a custom tactic might not be worth it.
